lib/fiddle/truffleruby.rb

```diff
@@ -0,0 +1,623 @@
+# frozen_string_literal: true
+# truffleruby_primitives: true
+
+# Copyright (c) 2019, 2024 Oracle and/or its affiliates. All rights reserved. This
+# code is released under a tri EPL/GPL/LGPL license. You can use it,
+# redistribute it and/or modify it under the terms of the:
+#
+# Eclipse Public License version 2.0, or
+# GNU General Public License version 2, or
+# GNU Lesser General Public License version 2.1.
+
+module Truffle::FiddleBackend
+
+  SIZEOF_CHAR   = Primitive.pointer_find_type_size(:char)
+  SIZEOF_SHORT  = Primitive.pointer_find_type_size(:short)
+  SIZEOF_INT    = Primitive.pointer_find_type_size(:int)
+  SIZEOF_LONG   = Primitive.pointer_find_type_size(:long)
+
+  CHAR_NFI_TYPE  = "SINT#{SIZEOF_CHAR * 8}"
+  UCHAR_NFI_TYPE  = "UINT#{SIZEOF_CHAR * 8}"
+  SHORT_NFI_TYPE  = "SINT#{SIZEOF_SHORT * 8}"
+  USHORT_NFI_TYPE  = "UINT#{SIZEOF_SHORT * 8}"
+  INT_NFI_TYPE  = "SINT#{SIZEOF_INT * 8}"
+  UINT_NFI_TYPE  = "UINT#{SIZEOF_INT * 8}"
+  LONG_NFI_TYPE = "SINT#{SIZEOF_LONG * 8}"
+  ULONG_NFI_TYPE = "UINT#{SIZEOF_LONG * 8}"
+
+  SIZE_T_TYPEDEF = Truffle::Config.lookup('platform.typedef.size_t')
```

Comment:
Everything under `Truffle::` is considered implementation details, with the exception of `Truffle::Interop`:
https://github.com/oracle/truffleruby/blob/master/doc/user/truffleruby-additions.md#unsupported-additional-functionality

It's a similar concern as with the Primitives, whatever is used in external gems needs to be made stable first.
Which might imply moving the method to maybe something like `TruffleRuby::FiddleSupport`, renaming the method or using a Primitive instead, etc.

Reply:
One trouble with this is e.g. if this is changed/moved/renamed then it won't work on TruffleRuby 24.1.0 and older versions.
Then we would need another solution for older versions.

For example a stable `Primitive` seems better here and exposing `Truffle::Config.lookup`.

Reply:
We can release a new version for newer TruffleRuby if it's needed.
Old TruffleRuby users can pin Fiddle gem to old version.

Reply:
> Old TruffleRuby users can pin Fiddle gem to old version.

The error for end users would be very unclear though (e.g. some NoMethodError), and they would likely have no idea how to solve this from the error message.
Also if for whatever reason it's a program run without Bundler then I don't think there is a way to pin the Fiddle version.

That's why I have always thought and still think the best solution is this PR for upcoming TruffleRuby versions, and using the stdlib for older TruffleRuby versions.
This is also what the FFI gem does for example:
https://github.com/ffi/ffi/blob/c128cede750242fe19945af8bd6c797728489ad5/lib/ffi.rb#L14-L27